x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,488 @@
+			// If half of the set visibilityTimeout passed and this is
+			// still ongoing, then change visibility timeout.
+			err := p.changeVisibilityTimeout(queueURL, visibilityTimeout, svcSQS, message.ReceiptHandle)
+			if err != nil {
+				p.logger.Error(errors.Wrap(err, "change message visibility failed"))
+			}
+			p.logger.Infof("Message visibility timeout updated to %v", visibilityTimeout)
+		}
+	}
+}
+
+func (p *Input) changeVisibilityTimeout(queueURL string, visibilityTimeout int64, svc *sqs.Client, receiptHandle *string) error {
+	req := svc.ChangeMessageVisibilityRequest(&sqs.ChangeMessageVisibilityInput{
+		QueueUrl:          &queueURL,
+		VisibilityTimeout: &visibilityTimeout,
+		ReceiptHandle:     receiptHandle,
+	})
+	_, err := req.Send(p.context)
+	return err
+}
+
+func getRegionFromQueueURL(queueURL string) (string, error) {
+	// get region from queueURL
+	// Example: https://sqs.us-east-1.amazonaws.com/627959692251/test-s3-logs
+	queueURLSplit := strings.Split(queueURL, ".")
+	if queueURLSplit[0] == "https://sqs" && queueURLSplit[2] == "amazonaws" {
+		return queueURLSplit[1], nil
+	}
+	return "", errors.New("queueURL is not in format: https://sqs.{REGION_ENDPOINT}.amazonaws.com/{ACCOUNT_NUMBER}/{QUEUE_NAME}")
+}
+
+// handle message
+func handleSQSMessage(m sqs.Message) ([]s3Info, error) {
+	msg := sqsMessage{}
+	err := json.Unmarshal([]byte(*m.Body), &msg)
+	if err != nil {
+		return nil, errors.Wrap(err, "json unmarshal sqs message body failed")
+	}
+
+	var s3Infos []s3Info
+	for _, record := range msg.Records {
+		if record.EventSource == "aws:s3" && record.EventName == "ObjectCreated:Put" {
+			s3Infos = append(s3Infos, s3Info{
+				region: record.AwsRegion,
+				name:   record.S3.bucket.Name,
+				key:    record.S3.object.Key,
+				arn:    record.S3.bucket.Arn,
+			})
+		}
+	}
+	return s3Infos, nil
+}
+
+func (p *Input) handleS3Objects(svc s3iface.ClientAPI, s3Infos []s3Info, errC chan error) error {
+	s3Context := &s3Context{
+		errC: errC,
+	}
+
+	for _, s3Info := range s3Infos {
+		objectHash := s3ObjectHash(s3Info)
+
+		// read from s3 object
+		reader, err := p.bufferedIORead(svc, s3Info)
+		if err != nil {
+			return errors.Wrap(err, "bufferedIORead failed")
+		}
+
+		offset := 0
+		for {
+			log, err := reader.ReadString('\n')
+			if log == "" {
+				break
+			}
+
+			if err != nil {
+				if err == io.EOF {
+					// create event for last line
+					offset += len([]byte(log))
+					event := createEvent(log, offset, s3Info, objectHash, s3Context)
+					err = p.forwardEvent(event)
+					if err != nil {
+						err = errors.Wrapf(err, "forwardEvent failed for %v", s3Info.key)
+						s3Context.Fail(err)
+						return err
+					}
+					return nil
+				}
+				return errors.Wrapf(err, "ReadString failed for %v", s3Info.key)
+			}
+
+			// create event per log line
+			offset += len([]byte(log))
+			event := createEvent(log, offset, s3Info, objectHash, s3Context)
+			err = p.forwardEvent(event)
```

Comment:
createEvent increases the ref count to 1 on the first message received. Now if a context switch happens, we have a chance of asynchronously setting the ref count to 0 before the next message is to be published. In this case errC will get a value and will be closed. Now this loop continues with the next message, increases the ref count back to 1 and the very next time we get an ACK setting the count back to 0 we have a panic, because we try sending to errC, which was closed already. This body of this function needs to start with a ref count of 1 and decrease it after until it has returned. All other processing will be asynchronously, saying you have no idea when you will get a result.

Reply:
good catch, I guess adding this at the beginning would do?:

```
s3Context.Inc()
defer s3Context.done()
```

Reply:
yes. This way the function body is an owner of s3Context for as long is required. Which is important, as the function body needs to pass ownership to `beat.Event` via createEvent. Rule of thumb about reference counting: one can only transfer (or add) ownership, if one is the actual owner.

Reply:
Thanks! Now it makes sense why in https://github.com/elastic/beats/pull/12640#discussion_r306086350 you initialized s3Context with `len(s3Infos)` and still did `s3Context.Inc()` afterwards. Thanks!